File: reascript/reabot_main.py
"""
ReaBot - Main Entry Point

This file is loaded by REAPER as a ReaScript. It owns the defer loop
and wires together all components:

  - DaemonClient: non-blocking TCP bridge to the Python daemon
  - ChatWindow: ReaImGui floating window
  - RenderPoller: waits for the async WAV render to complete
  - Extraction modules: track metadata, FX chain

To run: Actions -> Load ReaScript -> select this file
To stop: Run it again (toggle) or close the ReaBot window.

Requires:
  - ReaImGui installed via ReaPack
  - ReaBot daemon running (python -m daemon from the project root)
"""

import sys
import os
import json
import uuid

# Ensure the reascript package and its sibling dirs are importable
_SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
_PROJECT_ROOT = os.path.dirname(_SCRIPT_DIR)
if _PROJECT_ROOT not in sys.path:
    sys.path.insert(0, _PROJECT_ROOT)

# Also add the ReaTeam Extensions API path for imgui.py
_IMGUI_PATH = os.path.expanduser(
    "~/.config/REAPER/Scripts/ReaTeam Extensions/API"
)
if _IMGUI_PATH not in sys.path:
    sys.path.insert(0, _IMGUI_PATH)

import imgui
from reaper_python import RPR_defer, RPR_ShowConsoleMsg

from reascript.bridge.daemon_client import DaemonClient
from reascript.ui.chat_window import ChatWindow
from reascript.ui.theme import apply_theme, THEME_COLOR_COUNT, THEME_VAR_COUNT
from reascript.extraction.track_info import get_track_context
from reascript.extraction.fx_chain import get_selected_tracks_fx
from reascript.extraction.audio_render import RenderPoller, make_temp_path, cleanup_temp_file

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------
DAEMON_HOST = "127.0.0.1"
DAEMON_PORT = 9876

# ---------------------------------------------------------------------------
# Global state (lives across defer frames for the duration of the session)
# ---------------------------------------------------------------------------
_ctx = None           # ReaImGui context
_window = ChatWindow()
_client = DaemonClient(DAEMON_HOST, DAEMON_PORT)
_poller = RenderPoller()
_open = True          # Window open flag

# State for the current in-flight request
_pending_request_id: str | None = None
_pending_wav_path: str | None = None
_pending_prompt: str | None = None
_pending_stereo: bool = False
_streaming_msg_index: int = -1

# ---------------------------------------------------------------------------
# Daemon message handlers
# ---------------------------------------------------------------------------

def _on_response_chunk(msg: dict) -> None:
    """A token arrived from the LLM. Append to the streaming message."""
    global _streaming_msg_index
    if _streaming_msg_index >= 0:
        _window.append_chunk(_streaming_msg_index, msg.get("payload", {}).get("content", ""))


def _on_response_done(msg: dict) -> None:
    """LLM response is complete."""
    global _streaming_msg_index, _pending_request_id
    if _streaming_msg_index >= 0:
        _window.finish_streaming(_streaming_msg_index)
    _streaming_msg_index = -1
    _pending_request_id = None
    _window.set_status("Ready")
    # Clean up the temp WAV file
    if _pending_wav_path:
        cleanup_temp_file(_pending_wav_path)


def _on_analysis_result(msg: dict) -> None:
    """
    Raw analysis result (no LLM handler registered in daemon yet — Phase 4).
    Display a summary so the UI isn't blank during Phase 3.
    """
    global _streaming_msg_index, _pending_request_id
    analysis = msg.get("payload", {}).get("analysis", {})
    summary_parts = []
    if "rms_db" in analysis:
        summary_parts.append(f"RMS: {analysis['rms_db']:.1f} dB")
    if "spectral_centroid_hz" in analysis:
        summary_parts.append(f"Centroid: {analysis['spectral_centroid_hz']:.0f} Hz")
    if "musical" in analysis:
        m = analysis["musical"]
        summary_parts.append(f"BPM: {m.get('bpm', '?')} | Key: {m.get('key', '?')}")
    if "stereo" in analysis:
        s = analysis["stereo"]
        summary_parts.append(f"Width: {s.get('stereo_width', 0):.2f} | Compat: {s.get('mono_compatibility_score', 1):.2f}")

    if summary_parts:
        result_text = "Analysis complete (LLM not yet active):\n" + "\n".join(f"  {p}" for p in summary_parts)
    else:
        result_text = "Analysis complete. No metrics returned."

    if _streaming_msg_index >= 0:
        _window.append_chunk(_streaming_msg_index, result_text)
        _window.finish_streaming(_streaming_msg_index)

    _streaming_msg_index = -1
    _pending_request_id = None
    _window.set_status("Ready")
    if _pending_wav_path:
        cleanup_temp_file(_pending_wav_path)


def _on_error(msg: dict) -> None:
    """Daemon returned an error."""
    global _streaming_msg_index, _pending_request_id
    error = msg.get("payload", {}).get("error", "Unknown error")
    _window.add_system_message(f"Error: {error}")
    if _streaming_msg_index >= 0:
        _window.finish_streaming(_streaming_msg_index)
    _streaming_msg_index = -1
    _pending_request_id = None
    _window.set_status("Error")
    if _pending_wav_path:
        cleanup_temp_file(_pending_wav_path)


# Register handlers
_client.on("response_chunk",   _on_response_chunk)
_client.on("response_done",    _on_response_done)
_client.on("analysis_result",  _on_analysis_result)
_client.on("error",            _on_error)

# ---------------------------------------------------------------------------
# UI callbacks
# ---------------------------------------------------------------------------

def _handle_analyze(prompt: str, stereo: bool) -> None:
    """
    User clicked Analyze:
    1. Extract track metadata + FX chain.
    2. Start WAV render.
    3. Wait for render to complete (poller.tick()).
    4. Send analyze_track request to daemon.
    """
    global _pending_prompt, _pending_stereo, _pending_wav_path, _streaming_msg_index

    _pending_prompt = prompt
    _pending_stereo = stereo
    _pending_wav_path = make_temp_path()

    error = _poller.start(_pending_wav_path)
    if error:
        _window.add_system_message(f"Cannot render: {error}")
        _window.set_status("Ready")
        return

    _streaming_msg_index = _window.add_assistant_message()
    _window.set_status("Rendering...")


def _handle_chat(prompt: str) -> None:
    """User clicked Chat (no audio — just a follow-up question)."""
    global _streaming_msg_index, _pending_request_id

    request_id = str(uuid.uuid4())
    _pending_request_id = request_id
    _streaming_msg_index = _window.add_assistant_message()

    _client.send_message({
        "type": "chat",
        "id": request_id,
        "payload": {
            "user_message": prompt,
        }
    })
    _window.set_status("Waiting for response...")


_window.on_analyze = _handle_analyze
_window.on_chat = _handle_chat

# ---------------------------------------------------------------------------
# Setup ImGui context
# ---------------------------------------------------------------------------

def _setup() -> None:
    global _ctx
    _ctx = imgui.CreateContext("ReaBot")
    imgui.Attach(_ctx, imgui.GetBuiltinFont())
    apply_theme(_ctx)


_setup()

# ---------------------------------------------------------------------------
# Main defer loop
# ---------------------------------------------------------------------------

def _loop() -> None:
    global _open, _pending_wav_path, _pending_prompt, _pending_stereo
    global _streaming_msg_index, _pending_request_id

    if not _open:
        # Window was closed — clean up and stop the defer loop
        imgui.DestroyContext(_ctx)
        return

    # 1. Tick the daemon client (non-blocking socket I/O)
    was_connected = _client.is_connected
    _client.tick()
    now_connected = _client.is_connected

    if was_connected != now_connected:
        _window.set_connected(now_connected)
        if now_connected:
            _window.set_status("Ready")
        else:
            _window.set_status("Daemon not connected — start the daemon first")

    # 2. Tick the render poller
    if _poller.is_active:
        ready, render_error = _poller.tick()
        if render_error:
            _window.add_system_message(f"Render failed: {render_error}")
            _window.finish_streaming(_streaming_msg_index)
            _streaming_msg_index = -1
            _window.set_status("Ready")
            _pending_wav_path = None
        elif ready:
            # Render done — gather metadata and send to daemon
            _window.set_status("Analyzing...")
            track_context = get_track_context()
            fx_data = get_selected_tracks_fx()

            # Merge FX data into track context
            for i, track in enumerate(track_context.get("selected_tracks", [])):
                if i < len(fx_data):
                    track["fx_chain"] = fx_data[i].get("fx_chain", [])

            request_id = str(uuid.uuid4())
            _pending_request_id = request_id

            _client.send_message({
                "type": "analyze_track",
                "id": request_id,
                "payload": {
                    "wav_path": _pending_wav_path,
                    "track_metadata": track_context,
                    "user_question": _pending_prompt or "",
                    "stereo": _pending_stereo,
                }
            })

    # 3. Draw UI
    imgui.SetNextWindowPos(_ctx, 100, 100, imgui.Cond_FirstUseEver())
    _open = _window.draw(_ctx)

    # 4. Reschedule
    RPR_defer("_loop()")


# Start the loop
_loop()

File: reascript/reabot_main.lua
--[[
  ReaBot — reabot_main.lua
  Entry point. REAPER loads this file from the action list.

  Responsibilities:
  - Set up Lua require() paths relative to this script's location
  - Initialize all modules
  - Wire callbacks between modules
  - Own and drive the main defer loop
--]]

-- ── Path setup (allows require() to find sibling modules) ──
-- reaper.get_action_context() returns: is_valid, filename, sectionID, cmdID, mode, resolution, val
local _script_path   = ({ reaper.get_action_context() })[2]
local _script_dir    = _script_path:match("(.*[/\\])")

package.path         = _script_dir .. "?.lua;"
    .. _script_dir .. "?/init.lua;"
    .. package.path

-- ── Locate and add Mavriq LuaSockets / Batteries ──
local resource_path  = reaper.GetResourcePath()
local is_windows     = reaper.GetOS():match("Win")
local sep            = is_windows and "\\" or "/"

-- Path to Sockets package
local sockets_path   = table.concat(
  { resource_path, "Scripts", "Mavriq ReaScript Repository", "Various", "Mavriq-Lua-Sockets", "" }, sep)
-- Path to Batteries package
local batteries_path = table.concat(
  { resource_path, "Scripts", "Mavriq ReaScript Repository", "Various", "Mavriq-Lua-Batteries", "" }, sep)

-- Sockets package paths
package.path         = package.path .. ";" .. sockets_path .. "?.lua"
package.cpath        = package.cpath ..
    ";" .. sockets_path .. "?.so;" .. sockets_path .. "?.dll;" .. sockets_path .. "?.dylib"

-- Batteries package paths (fallback/alternative)
package.path         = package.path .. ";" .. batteries_path .. "lua" .. sep .. "?.lua"
package.cpath        = package.cpath ..
    ";" ..
    batteries_path ..
    "bin" .. sep .. "?.so;" .. batteries_path .. "bin" .. sep .. "?.dll;" .. batteries_path .. "bin" .. sep .. "?.dylib"

-- ── Load modules ─────────────────────────────────────────────────────────
local socket         = require("bridge.socket_client")
local render         = require("render.stem_render")
local track          = require("extraction.track")
local ui             = require("ui.chat")

-- ── Initialize ─────────────────────────────────────────────────────────────
ui.init()

local REABOT_MODEL_DISPLAY = ""
ui.set_model(REABOT_MODEL_DISPLAY)

socket.connect()

-- Route all daemon messages through the UI handler
socket.on_message(function(msg)
  local t = msg.type
  local p = msg.payload or {}

  -- status_ok: grab model name and enable Chat if we have stored history
  if t == "status_ok" then
    if p.model and p.model ~= "" then
      ui.set_model(p.model)
    end
    if p.has_history then
      ui.set_has_analysis(true)
      local turns = p.history_turns or 0
      if turns > 0 then
        ui.set_status(string.format("Loaded %d previous turn%s", turns, turns == 1 and "" or "s"))
      end
    end
  end

  -- WAV cleanup: both response_done and analysis_result complete the flow
  if t == "response_done" or t == "analysis_result" or t == "error" then
    _current_request_id = nil
    if _pending_wav then
      render.cleanup(_pending_wav)
      _pending_wav = nil
    end
  end

  ui.on_daemon_message(msg)
end)

-- ── Pending WAV path and current request ID (for deferred cleanup/cancel) ──
_pending_wav = nil -- global so the on_message closure above can reach it
_current_request_id = nil

-- ── ID generator (lightweight, no UUID lib needed) ────────────────────────
local _id_counter = 0
local function make_id()
  _id_counter = _id_counter + 1
  return string.format("reabot-%08x-%04d", os.time(), _id_counter)
end

-- ── Analyze button handler ─────────────────────────────────────────────────
ui.on_analyze_click = function(prompt, stereo)
  -- Guard: reconnect if we fell off
  if not socket.is_connected() then
    socket.reconnect()
    ui.show_error("Reconnecting to daemon — please try again in a moment.")
    return
  end

  -- Guard: need at least one selected track
  local target_track = track.get_first_track_handle()
  if not target_track then
    ui.show_error("No track selected. Select a track in REAPER before clicking Analyze.")
    return
  end

  -- Guard: render not already running
  if render.is_busy() then
    ui.show_error("A render is already in progress. Please wait.")
    return
  end

  -- Read track metadata (includes FX chain nested inside)
  local meta, err = track.get_first_selected()
  if not meta then
    ui.show_error(err or "Could not read track metadata.")
    return
  end

  -- Show user's message immediately so it feels responsive
  ui.add_user_message(prompt)
  ui.set_status("Rendering...")

  -- Start async render — callback fires when WAV file is ready
  render.start(target_track, function(wav_path)
    -- WAV is ready — send analysis request to daemon
    _pending_wav = wav_path
    ui.set_status("Analyzing...")

    local req_id = make_id()
    _current_request_id = req_id
    socket.send({
      type    = "analyze_track",
      id      = req_id,
      payload = {
        wav_path       = wav_path,
        track_metadata = meta, -- fx_chain is nested inside meta
        user_question  = prompt,
        stereo         = stereo,
        project_path   = reaper.GetProjectPath("") or "",
      },
    })
  end, function(render_err)
    -- Render failed
    ui.show_error("Render failed: " .. render_err)
    ui.set_status("Ready")
  end)
end

-- ── Chat button handler (follow-up, no render) ────────────────────────────
ui.on_chat_click = function(prompt)
  if not socket.is_connected() then
    ui.show_error("Not connected to daemon. Start the daemon and try again.")
    return
  end

  ui.add_user_message(prompt)
  ui.set_status("Thinking...")

  local req_id = make_id()
  _current_request_id = req_id
  socket.send({
    type    = "chat",
    id      = req_id,
    payload = {
      user_message = prompt,
    },
  })
end

-- ── Startup: clean up stale temp files from previous sessions ────────────
render.cleanup_stale(3600)

-- ── Send an initial STATUS ping — includes project_path so the daemon can
--    load stored conversation history────
socket.send({
  type    = "status",
  id      = make_id(),
  payload = { project_path = reaper.GetProjectPath("") or "" },
})

-- ── Main defer loop ────────────────────────────────────────────────────────
local function loop()
  socket.tick()
  render.tick()

  -- Sync connection state to UI indicator every frame
  ui.set_connected(socket.is_connected())

  -- Draw UI — returns false when user clicks the X button
  local keep_open = ui.draw()
  if keep_open then
    reaper.defer(loop)
  else
    -- Window closed
    if _current_request_id and socket.is_connected() then
      socket.send({
        type    = "cancel",
        id      = make_id(),
        payload = { target_id = _current_request_id },
      })
      -- Tick once more to flush the send queue exiting
      socket.tick()
    end

    if _pending_wav then
      render.cleanup(_pending_wav)
      _pending_wav = nil
    end
    reaper.ShowConsoleMsg("[ReaBot] Window closed.\n")
  end
end

-- Kick off the loop
loop()
